### PR TITLE
enha: make getThrowable public and improve set contexts

### PR DESCRIPTION
**Pull request by marandaneto**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [X] Enhancement
- [ ] Refactoring

## :scroll: Description
enha: make getThrowable public and improve set contexts

## :bulb: Motivation and Context
`getThrowable` is useful to use `instanceof` for filtering.
`setContexts` now have overloads for the most common data types as they need to be an object.

## :green_heart: How did you test it?
running it

## :pencil: Checklist

- [X] I reviewed submitted code
- [ ] I added tests to verify changes
- [X] All tests passing
- [X] No breaking changes

### CODE REVIEW

File: sentry/src/main/java/io/sentry/Scope.java [hunk 1]
```diff
@@ -336,6 +337,42 @@ public void setContexts(final @NotNull String key, final @NotNull Object value)
...
 
+  /**
+   * Sets the Scope's contexts
+   *
+   * @param key the context key
+   * @param value the context value
+   */
+  public void setContexts(final @NotNull String key, final @NotNull Boolean value) {
+    final Map<String, Boolean> map = new HashMap<>();
+    map.put("value", value);
```
Comment (philipphofmann) [inline sentry/src/main/java/io/sentry/Scope.java:348]: `h`: Why not just directly pass the value and key to `setContexts` and wrap it in a HashMap? We don't do this in [cocoa](https://github.com/getsentry/sentry-cocoa/blob/980d89d748992c1dcac6994eced3c074ff765917/Sources/Sentry/SentryScope.m#L149-L154) and [javascript](https://github.com/getsentry/sentry-javascript/blob/6c33bfcd27bb45cfbdc2459167c3b305ef93148c/packages/hub/src/scope.ts#L188-L198)

File: sentry/src/main/java/io/sentry/SentryEvent.java [hunk 2]
```diff
@@ -14,7 +14,10 @@
 public final class SentryEvent implements IUnknownPropertiesConsumer {
   private SentryId eventId;
   private final Date timestamp;
-  private transient Throwable throwable;
+
+  /** The captured Throwable */
```
Comment (philipphofmann) [inline sentry/src/main/java/io/sentry/SentryEvent.java:18]: `m`: I think this change could be in another PR. It also nicer for the changelog IMO, but the change itself looks good. 👍

**Review by philipphofmann — COMMENTED** [review]
Nice that we fix those things 👍

File: buildSrc/src/main/java/Config.kt [hunk 3]
```diff
@@ -9,7 +9,7 @@ object Config {
     val springKotlinCompatibleLanguageVersion = "1.3"
 
     object BuildPlugins {
-        val androidGradle = "com.android.tools.build:gradle:4.0.1"
+        val androidGradle = "com.android.tools.build:gradle:4.0.2"
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply buildSrc/src/main/java/Config.kt:12]: the IDE changed it automatically and I just kept it :( I always do Bump PRs

File: sentry/src/main/java/io/sentry/Scope.java (hunk 1 above)
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/Scope.java:348]: @philipphofmann because your value is already a `NSDictionary`, set `setContexts` doesn't accept primitive types, but only Objects, otherwise the backend will complain.
so we wrap it for the common data types, but one is still able to pass a serializable object if they want to.
so if one does: setContext("a", "b")
it should be:

```
"a": {
value: "b"
}
```

and not:

`"a": "b"`

File: sentry/src/main/java/io/sentry/SentryEvent.java (hunk 2 above)
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/SentryEvent.java:18]: but this is the title of the PR: `enha: make getThrowable public and improve set contexts`
I know its a mix of 2 changes, but both changes unblock this https://github.com/getsentry/sentry-docs/pull/2460 so it made sense to be done together, happy to split them if you think thats important.

File: buildSrc/src/main/java/Config.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (marandaneto) [reply buildSrc/src/main/java/Config.kt:12]: https://github.com/getsentry/sentry-java/pull/968

File: sentry/src/main/java/io/sentry/Scope.java (hunk 1 above)
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/Scope.java:348]: if we change the main method setContexts(key, value (as a Map), people wont be able to do a setContexts(key, theirOwnInstanceOfASerializableClass)

File: sentry/src/main/java/io/sentry/SentryEvent.java (hunk 2 above)
Comment (philipphofmann) [reply sentry/src/main/java/io/sentry/SentryEvent.java:18]: I don't have a strong opinion on this, just thought it would be cleaner.

File: sentry/src/main/java/io/sentry/Scope.java (hunk 1 above)
Comment (philipphofmann) [reply sentry/src/main/java/io/sentry/Scope.java:348]: Ah, now I think I get your point. These are convenience methods, right? We can already do 
```
HashMap<String, Object> map = new HashMap<>();
map.put("name", "FireFighter");
map.put("age", 19);
Sentry.configureScope(scope -> scope.setContexts("character", map));
```
We don't have them on Cocoa. Why did you add them? Should I also add them in Cocoa?

Comment (marandaneto) [reply sentry/src/main/java/io/sentry/Scope.java:348]: @philipphofmann it'd be nice to have those convenience methods, yes, but right now on Java it's kinda buggy.
`Contexts extends ConcurrentHashMap<String, Object>`

so one already could do setCotexts(x, y) but if `y` is a primitive type, it was not gonna be wrapped in an object like a `map` or so, so it'd trigger an error on the backend, every object on `contexts` should not be primitive, but an `object` but as in Java everything extends from an Object, the API allowed them to make a mistake.

File: sentry/src/main/java/io/sentry/SentryEvent.java (hunk 2 above)
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/SentryEvent.java:18]: ok, i will keep as it is now because they are related to the docs, but next time i will split them up, thanks

File: sentry/src/main/java/io/sentry/Scope.java (hunk 1 above)
Comment (philipphofmann) [reply sentry/src/main/java/io/sentry/Scope.java:348]: In Cocoa, we have `NSMutableDictionary<NSString *, NSDictionary<NSString *, id> *> *`, which allows you to put anything into the dict. We have
https://github.com/getsentry/sentry-cocoa/blob/master/Sources/Sentry/NSDictionary+SentrySanitize.m to make it work. Maybe something like this could help us also in Java?

Comment (marandaneto) [reply sentry/src/main/java/io/sentry/Scope.java:348]: @philipphofmann this is already done automatically by Gson.
what I mean is, this still forces people to pass a Map, it's verbose IMO

**Review by bruno-garcia — APPROVED** [review]

File: sentry/src/main/java/io/sentry/Scope.java (hunk 1 above)
Comment (philipphofmann) [reply sentry/src/main/java/io/sentry/Scope.java:348]: That is true indeed. Thanks for the explanation.

**Comment by maciejwalkowiak** [comment]
❤️ this change. Thanks @marandaneto!
